File: resources/android/src/com/victorycodedev/plugins/toastkit/bridge/ToastKitBridgeNormalizer.kt
package com.victorycodedev.plugins.toastkit.bridge

import androidx.compose.ui.graphics.Color
import com.victorycodedev.plugins.toastkit.model.*
import org.json.JSONObject

internal class ToastKitInputException(message: String) : IllegalArgumentException(message)

internal object ToastKitBridgeNormalizer {
    fun show(parameters: Map<String, Any>): ToastKitConfiguration {
        val version = (parameters["contract_version"] as? Number)?.toInt() ?: 1
        requireInput(version == 1, "Unsupported contract_version: $version")
        return configuration(parameters)
    }

    fun update(parameters: Map<String, Any>): Pair<String, Map<String, Any>> {
        val id = requiredString(parameters, "id")
        val changes = stringMap(parameters["changes"]) ?: throw ToastKitInputException("changes must be an object")
        requireInput(changes.isNotEmpty(), "changes must not be empty")
        return id to changes
    }

    fun updateUnique(parameters: Map<String, Any>): Pair<String, Map<String, Any>> {
        val key = requiredString(parameters, "unique_key")
        val changes = stringMap(parameters["changes"]) ?: throw ToastKitInputException("changes must be an object")
        requireInput(changes.isNotEmpty(), "changes must not be empty")
        return key to changes
    }

    fun id(parameters: Map<String, Any>): String = requiredString(parameters, "id")
    fun uniqueKey(parameters: Map<String, Any>): String = requiredString(parameters, "unique_key")

    fun applyChanges(current: ToastKitConfiguration, changes: Map<String, Any>): ToastKitConfiguration {
        var next = current
        changes["message"]?.let { next = next.copy(message = nonEmptyString(it, "message")) }
        if (changes.containsKey("title")) next = next.copy(title = nullableString(changes["title"], "title"))
        changes["variant"]?.let {
            val name = variant(it)
            next = next.copy(variant = name, style = variantStyle(name), icon = defaultIcon(name))
        }
        if (changes.containsKey("icon")) next = next.copy(icon = optionalValue(changes["icon"])?.let(::icon))
        changes["position"]?.let { next = next.copy(position = position(it)) }
        changes["animation"]?.let { next = next.copy(animation = animation(it)) }
        changes["direction"]?.let { next = next.copy(direction = direction(it)) }
        changes["native"]?.let {
            val mode = nativeMode(it)
            next = next.copy(nativeIos = mode.first, nativeAndroid = mode.second)
        }
        if (changes.containsKey("progress")) next = next.copy(progress = optionalValue(changes["progress"])?.let(::progress))
        changes["loading"]?.let { next = next.copy(loading = boolean(it, "loading")) }
        changes["swipe_to_dismiss"]?.let { next = next.copy(swipeToDismiss = boolean(it, "swipe_to_dismiss")) }
        changes["dismissible"]?.let { next = next.copy(dismissible = boolean(it, "dismissible")) }
        if (changes.containsKey("action")) next = next.copy(action = optionalValue(changes["action"])?.let(::action))
        changes["strategy"]?.let { next = next.copy(strategy = strategy(it)) }
        changes["max_visible"]?.let { next = next.copy(maxVisible = positiveInt(it, "max_visible")) }
        changes["style"]?.let { next = next.copy(style = style(it, next.style)) }
        if (changes.containsKey("text")) next = next.copy(text = mergeTypography(next.text, changes["text"], "text"))
        if (changes.containsKey("title_text")) next = next.copy(titleText = mergeTypography(next.titleText, changes["title_text"], "title_text"))

        val persistentChanged = changes.containsKey("persistent")
        val durationChanged = changes.containsKey("duration")
        if (persistentChanged || durationChanged) {
            val persistent = changes["persistent"]?.let { boolean(it, "persistent") } ?: false
            val duration = optionalValue(changes["duration"])
            next = next.copy(durationMs = if (persistent) null else duration?.let { positiveLong(it, "duration") } ?: 3000L)
        }
        return next
    }

    private fun configuration(values: Map<String, Any>): ToastKitConfiguration {
        val persistent = values["persistent"]?.let { boolean(it, "persistent") } ?: false
        return ToastKitConfiguration(
            id = requiredString(values, "id"),
            uniqueKey = nullableString(values["unique_key"], "unique_key"),
            nativeIos = nativeMode(values["native"] ?: emptyMap<String, Any>()).first,
            nativeAndroid = nativeMode(values["native"] ?: emptyMap<String, Any>()).second,
            message = requiredString(values, "message"),
            title = nullableString(values["title"], "title"),
            variant = variant(values["variant"] ?: "neutral"),
            icon = optionalValue(values["icon"])?.let(::icon),
            position = position(values["position"] ?: "bottom"),
            durationMs = if (persistent) null else positiveLong(values["duration"] ?: 3000, "duration"),
            animation = animation(values["animation"] ?: "scale"),
            direction = direction(values["direction"] ?: "auto"),
            progress = optionalValue(values["progress"])?.let(::progress),
            loading = values["loading"]?.let { boolean(it, "loading") } ?: false,
            swipeToDismiss = values["swipe_to_dismiss"]?.let { boolean(it, "swipe_to_dismiss") } ?: true,
            dismissible = values["dismissible"]?.let { boolean(it, "dismissible") } ?: false,
            action = optionalValue(values["action"])?.let(::action),
            style = style(values["style"] ?: emptyMap<String, Any>(), variantStyle(values["variant"] as? String ?: "neutral")),
            strategy = strategy(values["strategy"] ?: "queue"),
            maxVisible = positiveInt(values["max_visible"] ?: 3, "max_visible"),
            text = optionalValue(values["text"])?.let(::typography),
            titleText = optionalValue(values["title_text"])?.let(::typography),
        )
    }

    private fun icon(value: Any): ToastKitIcon {
        val map = stringMap(value) ?: throw ToastKitInputException("icon must be an object")
        val name = nullableString(map["name"], "icon.name")
        val android = nullableString(map["android"], "icon.android")
        requireInput(name != null || android != null, "icon requires name or android")
        return ToastKitIcon(name, android)
    }

    private fun action(value: Any): ToastKitAction {
        val map = stringMap(value) ?: throw ToastKitInputException("action must be an object")
        return ToastKitAction(requiredString(map, "id"), requiredString(map, "label"))
    }

    private fun nativeMode(value: Any): Pair<Boolean, Boolean> {
        val map = stringMap(value) ?: throw ToastKitInputException("native must be an object")
        return (map["ios"]?.let { boolean(it, "native.ios") } ?: false) to
            (map["android"]?.let { boolean(it, "native.android") } ?: false)
    }

    private fun style(value: Any, fallback: ToastKitStyle): ToastKitStyle {
        val map = stringMap(value) ?: throw ToastKitInputException("style must be an object")
        return fallback.copy(
            background = map["background"]?.let { color(it, "background") } ?: fallback.background,
            foreground = map["foreground"]?.let { color(it, "foreground") } ?: fallback.foreground,
            iconColor = map["icon_color"]?.let { color(it, "icon_color") } ?: fallback.iconColor,
            actionColor = map["action_color"]?.let { color(it, "action_color") } ?: fallback.actionColor,
            cornerRadius = map["corner_radius"]?.let { nonNegativeFloat(it, "corner_radius") } ?: fallback.cornerRadius,
            padding = map["padding"]?.let { nonNegativeFloat(it, "padding") } ?: fallback.padding,
            shadow = map["shadow"]?.let { boolean(it, "shadow") } ?: fallback.shadow,
        )
    }

    private fun variantStyle(variant: String): ToastKitStyle {
        val colors = when (variant) {
            "success" -> listOf("#166534", "#FFFFFF", "#86EFAC", "#BBF7D0")
            "error" -> listOf("#991B1B", "#FFFFFF", "#FCA5A5", "#FECACA")
            "warning" -> listOf("#92400E", "#FFFFFF", "#FDE68A", "#FEF3C7")
            "info" -> listOf("#1E40AF", "#FFFFFF", "#93C5FD", "#BFDBFE")
            else -> listOf("#1F2937", "#FFFFFF", "#D1D5DB", "#E5E7EB")
        }
        return ToastKitStyle(parseColor(colors[0]), parseColor(colors[1]), parseColor(colors[2]), parseColor(colors[3]), 16f, 16f, true)
    }

    private fun defaultIcon(variant: String): ToastKitIcon? = when (variant) {
        "success" -> ToastKitIcon("check", null)
        "error" -> ToastKitIcon("error", null)
        "warning" -> ToastKitIcon("warning", null)
        "info" -> ToastKitIcon("info", null)
        else -> null
    }

    private fun typography(value: Any): ToastKitTypography {
        val map = stringMap(value) ?: throw ToastKitInputException("typography must be an object")
        if (map.isEmpty()) throw ToastKitInputException("typography must not be empty")
        return ToastKitTypography(
            font = map["font"]?.let { nonEmptyString(it, "font") },
            size = map["size"]?.let { textSize(it) },
            weight = map["weight"]?.let { textWeight(it) },
            align = map["align"]?.let { textAlign(it) },
            italic = map["italic"]?.let { boolean(it, "italic") },
        )
    }

    private fun mergeTypography(current: ToastKitTypography?, value: Any?, name: String): ToastKitTypography? {
        if (isNull(value)) return current
        val map = stringMap(value) ?: throw ToastKitInputException("$name must be an object")
        if (map.isEmpty()) return current
        val base = current ?: ToastKitTypography()
        return base.copy(
            font = if (map.containsKey("font")) map["font"]?.let { nonEmptyString(it, "font") } else base.font,
            size = if (map.containsKey("size")) map["size"]?.let { textSize(it) } else base.size,
            weight = if (map.containsKey("weight")) map["weight"]?.let { textWeight(it) } else base.weight,
            align = if (map.containsKey("align")) map["align"]?.let { textAlign(it) } else base.align,
            italic = if (map.containsKey("italic")) map["italic"]?.let { boolean(it, "italic") } else base.italic,
        )
    }

    private fun color(value: Any, name: String): Color {
        val text = value as? String ?: throw ToastKitInputException("$name must be a color string")
        return try { parseColor(text) } catch (_: IllegalArgumentException) { throw ToastKitInputException("Invalid $name color") }
    }

    private fun parseColor(value: String): Color {
        val hex = value.removePrefix("#")
        require(hex.length == 3 || hex.length == 6 || hex.length == 8)
        val expanded = if (hex.length == 3) hex.map { "$it$it" }.joinToString("") else hex
        val argb = if (expanded.length == 6) "FF$expanded" else expanded
        return Color(argb.toLong(16))
    }

    private fun variant(value: Any): String = enumString(value, "variant", setOf("neutral", "success", "error", "warning", "info"))
    private fun position(value: Any) = ToastKitPosition.valueOf(enumString(value, "position", setOf("top", "center", "bottom")).uppercase())
    private fun animation(value: Any) = ToastKitAnimation.valueOf(enumString(value, "animation", setOf("fade", "slide", "scale", "spring", "snap", "pop", "reveal", "bounce")).uppercase())
    private fun direction(value: Any) = ToastKitDirection.valueOf(enumString(value, "direction", setOf("auto", "left", "right", "top", "bottom")).uppercase())
    private fun strategy(value: Any) = ToastKitStrategy.valueOf(enumString(value, "strategy", setOf("queue", "stack")).uppercase())
    private fun textSize(value: Any) = ToastKitTextSize.valueOf(enumString(value, "text size", setOf("xs", "sm", "base", "lg", "xl")).uppercase())
    private fun textWeight(value: Any) = ToastKitTextWeight.valueOf(enumString(value, "text weight", setOf("normal", "medium", "semibold", "bold")).uppercase())
    private fun textAlign(value: Any) = ToastKitTextAlign.valueOf(enumString(value, "text align", setOf("left", "center", "right")).uppercase())
    private fun enumString(value: Any, name: String, allowed: Set<String>): String {
        val text = nonEmptyString(value, name).lowercase()
        requireInput(text in allowed, "Invalid $name: $text")
        return text
    }
    private fun requiredString(map: Map<String, Any>, key: String) = nonEmptyString(map[key], key)
    private fun nonEmptyString(value: Any?, name: String): String {
        val text = value as? String ?: throw ToastKitInputException("$name must be a string")
        requireInput(text.isNotBlank(), "$name must not be empty")
        return text
    }
    private fun nullableString(value: Any?, name: String): String? {
        if (isNull(value)) return null
        return nonEmptyString(value, name)
    }
    private fun optionalValue(value: Any?): Any? = if (isNull(value)) null else value
    private fun isNull(value: Any?): Boolean = value == null || value === JSONObject.NULL
    private fun boolean(value: Any, name: String) = value as? Boolean ?: throw ToastKitInputException("$name must be boolean")
    private fun positiveLong(value: Any, name: String): Long {
        val result = (value as? Number)?.toLong() ?: throw ToastKitInputException("$name must be numeric")
        requireInput(result > 0, "$name must be greater than zero")
        return result
    }
    private fun positiveInt(value: Any, name: String) = positiveLong(value, name).toInt()
    private fun nonNegativeFloat(value: Any, name: String): Float {
        val result = (value as? Number)?.toFloat() ?: throw ToastKitInputException("$name must be numeric")
        requireInput(result >= 0, "$name must not be negative")
        return result
    }
    private fun progress(value: Any): Float {
        val result = (value as? Number)?.toFloat() ?: throw ToastKitInputException("progress must be numeric")
        requireInput(result.isFinite(), "progress must be finite")
        return result.coerceIn(0f, 100f)
    }
    private fun stringMap(value: Any?): Map<String, Any>? = when (value) {
        is JSONObject -> value.keys().asSequence().associate { key -> key to value.get(key) }
        is Map<*, *> -> value.entries.associate { (key, v) -> key.toString() to (v ?: JSONObject.NULL) }
        else -> null
    }
    private fun requireInput(condition: Boolean, message: String) { if (!condition) throw ToastKitInputException(message) }
}

File: resources/ios/Sources/Bridge/ToastKitBridgeNormalizer.swift
import SwiftUI

enum ToastKitInputError: LocalizedError {
    case invalid(String)
    var errorDescription: String? { if case .invalid(let message) = self { message } else { "Invalid ToastKit payload" } }
}

enum ToastKitBridgeNormalizer {
    static func show(_ values: [String: Any]) throws -> ToastKitConfiguration {
        let version = (values["contract_version"] as? NSNumber)?.intValue ?? 1
        guard version == 1 else { throw ToastKitInputError.invalid("Unsupported contract_version: \(version)") }
        return try configuration(values)
    }

    static func update(_ values: [String: Any]) throws -> (String, [String: Any]) {
        let id = try requiredString(values, "id")
        guard let changes = values["changes"] as? [String: Any], !changes.isEmpty else {
            throw ToastKitInputError.invalid("changes must be a non-empty object")
        }
        return (id, changes)
    }

    static func updateUnique(_ values: [String: Any]) throws -> (String, [String: Any]) {
        let key = try requiredString(values, "unique_key")
        guard let changes = values["changes"] as? [String: Any], !changes.isEmpty else {
            throw ToastKitInputError.invalid("changes must be a non-empty object")
        }
        return (key, changes)
    }

    static func id(_ values: [String: Any]) throws -> String { try requiredString(values, "id") }
    static func uniqueKey(_ values: [String: Any]) throws -> String { try requiredString(values, "unique_key") }

    static func applying(_ changes: [String: Any], to current: ToastKitConfiguration) throws -> ToastKitConfiguration {
        var next = current
        if let value = changes["message"] { next.message = try nonEmptyString(value, "message") }
        if changes.keys.contains("title") { next.title = try nullableString(changes["title"], "title") }
        if let value = changes["variant"] {
            let name = try variant(value)
            next.variant = name
            next.style = variantStyle(name)
            next.icon = defaultIcon(name)
        }
        if let value = changes["icon"] { next.icon = try icon(value) }
        if let value = changes["position"] { next.position = try enumValue(value, "position") }
        if let value = changes["animation"] { next.animation = try enumValue(value, "animation") }
        if let value = changes["direction"] { next.direction = try enumValue(value, "direction") }
        if changes.keys.contains("progress") { next.progress = try changes["progress"].map(progress) }
        if let value = changes["loading"] { next.loading = try boolean(value, "loading") }
        if let value = changes["swipe_to_dismiss"] { next.swipeToDismiss = try boolean(value, "swipe_to_dismiss") }
        if let value = changes["dismissible"] { next.dismissible = try boolean(value, "dismissible") }
        if changes.keys.contains("action") { next.action = try changes["action"].map(action) }
        if let value = changes["strategy"] { next.strategy = try enumValue(value, "strategy") }
        if let value = changes["max_visible"] { next.maxVisible = try positiveInt(value, "max_visible") }
        if let value = changes["style"] { next.style = try style(value, fallback: next.style) }
        if changes.keys.contains("text") { next.text = try mergeTypography(next.text, changes["text"], "text") }
        if changes.keys.contains("title_text") { next.titleText = try mergeTypography(next.titleText, changes["title_text"], "title_text") }

        if changes.keys.contains("duration") || changes.keys.contains("persistent") {
            let persistent = try changes["persistent"].map { try boolean($0, "persistent") } ?? false
            next.durationMilliseconds = persistent ? nil : try changes["duration"].map { try positiveInt($0, "duration") } ?? 3000
        }
        return next
    }

    private static func configuration(_ values: [String: Any]) throws -> ToastKitConfiguration {
        let variantName = try variant(values["variant"] ?? "neutral")
        let persistent = try values["persistent"].map { try boolean($0, "persistent") } ?? false
        return ToastKitConfiguration(
            id: try requiredString(values, "id"),
            uniqueKey: try nullableString(values["unique_key"], "unique_key"),
            message: try requiredString(values, "message"),
            title: try nullableString(values["title"], "title"),
            variant: variantName,
            icon: try values["icon"].map(icon),
            position: try enumValue(values["position"] ?? "bottom", "position"),
            durationMilliseconds: persistent ? nil : try positiveInt(values["duration"] ?? 3000, "duration"),
            animation: try enumValue(values["animation"] ?? "scale", "animation"),
            direction: try enumValue(values["direction"] ?? "auto", "direction"),
            progress: try values["progress"].map(progress),
            loading: try values["loading"].map { try boolean($0, "loading") } ?? false,
            swipeToDismiss: try values["swipe_to_dismiss"].map { try boolean($0, "swipe_to_dismiss") } ?? true,
            dismissible: try values["dismissible"].map { try boolean($0, "dismissible") } ?? false,
            action: try values["action"].map(action),
            style: try style(values["style"] ?? [String: Any](), fallback: variantStyle(variantName)),
            strategy: try enumValue(values["strategy"] ?? "queue", "strategy"),
            maxVisible: try positiveInt(values["max_visible"] ?? 3, "max_visible"),
            text: try values["text"].map(typography),
            titleText: try values["title_text"].map(typography)
        )
    }

    private static func icon(_ value: Any) throws -> ToastKitIconConfiguration {
        guard let map = value as? [String: Any] else { throw ToastKitInputError.invalid("icon must be an object") }
        let name = try nullableString(map["name"], "icon.name")
        let ios = try nullableString(map["ios"], "icon.ios")
        guard name != nil || ios != nil else { throw ToastKitInputError.invalid("icon requires name or ios") }
        return ToastKitIconConfiguration(name: name, ios: ios)
    }

    private static func action(_ value: Any) throws -> ToastKitActionConfiguration {
        guard let map = value as? [String: Any] else { throw ToastKitInputError.invalid("action must be an object") }
        return ToastKitActionConfiguration(id: try requiredString(map, "id"), label: try requiredString(map, "label"))
    }

    private static func typography(_ value: Any) throws -> ToastKitTypographyConfiguration {
        guard let map = value as? [String: Any] else { throw ToastKitInputError.invalid("typography must be an object") }
        guard !map.isEmpty else { throw ToastKitInputError.invalid("typography must not be empty") }
        var result = ToastKitTypographyConfiguration(font: nil, size: nil, weight: nil, align: nil, italic: nil)
        if let v = map["font"] { result.font = try nonEmptyString(v, "font") }
        if let v = map["size"] { result.size = try enumValue(v, "text size") }
        if let v = map["weight"] { result.weight = try enumValue(v, "text weight") }
        if let v = map["align"] { result.align = try enumValue(v, "text align") }
        if let v = map["italic"] { result.italic = try boolean(v, "italic") }
        return result
    }

    private static func mergeTypography(_ current: ToastKitTypographyConfiguration?, _ value: Any?, _ name: String) throws -> ToastKitTypographyConfiguration? {
        guard value != nil, !(value is NSNull) else { return current }
        guard let map = value as? [String: Any] else { throw ToastKitInputError.invalid("\(name) must be an object") }
        guard !map.isEmpty else { return current }
        var next = current ?? ToastKitTypographyConfiguration(font: nil, size: nil, weight: nil, align: nil, italic: nil)
        if let value = map["font"] { next.font = try nonEmptyString(value, "font") }
        if let value = map["size"] { next.size = try enumValue(value, "text size") }
        if let value = map["weight"] { next.weight = try enumValue(value, "text weight") }
        if let value = map["align"] { next.align = try enumValue(value, "text align") }
        if let value = map["italic"] { next.italic = try boolean(value, "italic") }
        return next
    }

    private static func style(_ value: Any, fallback: ToastKitStyleConfiguration) throws -> ToastKitStyleConfiguration {
        guard let map = value as? [String: Any] else { throw ToastKitInputError.invalid("style must be an object") }
        var result = fallback
        if let value = map["background"] { result.background = try color(value, "background") }
        if let value = map["foreground"] { result.foreground = try color(value, "foreground") }
        if let value = map["icon_color"] { result.iconColor = try color(value, "icon_color") }
        if let value = map["action_color"] { result.actionColor = try color(value, "action_color") }
        if let value = map["corner_radius"] { result.cornerRadius = try nonNegative(value, "corner_radius") }
        if let value = map["padding"] { result.padding = try nonNegative(value, "padding") }
        if let value = map["shadow"] { result.shadow = try boolean(value, "shadow") }
        return result
    }

    private static func variantStyle(_ variant: String) -> ToastKitStyleConfiguration {
        let colors: [String] = switch variant {
        case "success": ["#166534", "#FFFFFF", "#86EFAC", "#BBF7D0"]
        case "error": ["#991B1B", "#FFFFFF", "#FCA5A5", "#FECACA"]
        case "warning": ["#92400E", "#FFFFFF", "#FDE68A", "#FEF3C7"]
        case "info": ["#1E40AF", "#FFFFFF", "#93C5FD", "#BFDBFE"]
        default: ["#1F2937", "#FFFFFF", "#D1D5DB", "#E5E7EB"]
        }
        return ToastKitStyleConfiguration(background: safeColor(colors[0]), foreground: safeColor(colors[1]), iconColor: safeColor(colors[2]), actionColor: safeColor(colors[3]), cornerRadius: 16, padding: 16, shadow: true)
    }

    private static func defaultIcon(_ variant: String) -> ToastKitIconConfiguration? {
        switch variant {
        case "success": ToastKitIconConfiguration(name: "check", ios: nil)
        case "error": ToastKitIconConfiguration(name: "error", ios: nil)
        case "warning": ToastKitIconConfiguration(name: "warning", ios: nil)
        case "info": ToastKitIconConfiguration(name: "info", ios: nil)
        default: nil
        }
    }

    private static func color(_ value: Any, _ name: String) throws -> Color {
        guard let text = value as? String else { throw ToastKitInputError.invalid("\(name) must be a color string") }
        do { return try parseColor(text) } catch { throw ToastKitInputError.invalid("Invalid \(name) color") }
    }

    private static func parseColor(_ value: String) throws -> Color {
        var hex = value.hasPrefix("#") ? String(value.dropFirst()) : value
        guard [3, 6, 8].contains(hex.count), hex.allSatisfy({ $0.isHexDigit }) else { throw ToastKitInputError.invalid("Invalid color") }
        if hex.count == 3 { hex = hex.map { "\($0)\($0)" }.joined() }
        guard let number = UInt64(hex, radix: 16) else { throw ToastKitInputError.invalid("Invalid color") }
        let a, r, g, b: Double
        if hex.count == 8 { a = Double((number >> 24) & 255); r = Double((number >> 16) & 255); g = Double((number >> 8) & 255); b = Double(number & 255) }
        else { a = 255; r = Double((number >> 16) & 255); g = Double((number >> 8) & 255); b = Double(number & 255) }
        return Color(red: r / 255, green: g / 255, blue: b / 255, opacity: a / 255)
    }

    private static func safeColor(_ value: String) -> Color { (try? parseColor(value)) ?? .black }

    private static func variant(_ value: Any) throws -> String {
        let text = try nonEmptyString(value, "variant").lowercased()
        guard ["neutral", "success", "error", "warning", "info"].contains(text) else { throw ToastKitInputError.invalid("Invalid variant: \(text)") }
        return text
    }
    private static func enumValue<T: RawRepresentable>(_ value: Any, _ name: String) throws -> T where T.RawValue == String {
        let text = try nonEmptyString(value, name).lowercased()
        guard let result = T(rawValue: text) else { throw ToastKitInputError.invalid("Invalid \(name): \(text)") }
        return result
    }
    private static func requiredString(_ map: [String: Any], _ key: String) throws -> String { try nonEmptyString(map[key], key) }
    private static func nonEmptyString(_ value: Any?, _ name: String) throws -> String {
        guard let text = value as? String, !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { throw ToastKitInputError.invalid("\(name) must be a non-empty string") }
        return text
    }
    private static func nullableString(_ value: Any?, _ name: String) throws -> String? { value == nil || value is NSNull ? nil : try nonEmptyString(value, name) }
    private static func boolean(_ value: Any, _ name: String) throws -> Bool {
        guard let result = value as? Bool else { throw ToastKitInputError.invalid("\(name) must be boolean") }; return result
    }
    private static func positiveInt(_ value: Any, _ name: String) throws -> Int {
        guard let result = (value as? NSNumber)?.intValue, result > 0 else { throw ToastKitInputError.invalid("\(name) must be greater than zero") }; return result
    }
    private static func nonNegative(_ value: Any, _ name: String) throws -> CGFloat {
        guard let result = (value as? NSNumber)?.doubleValue, result >= 0 else { throw ToastKitInputError.invalid("\(name) must not be negative") }; return CGFloat(result)
    }
    private static func progress(_ value: Any) throws -> Double {
        guard let result = (value as? NSNumber)?.doubleValue, result.isFinite else {
            throw ToastKitInputError.invalid("progress must be a finite number")
        }
        return min(100, max(0, result))
    }
}
